types/src/database_sql.rs

```diff
@@ -226,6 +225,83 @@ impl Database {
         }))
     }
 
+    pub async fn upsert_collection(
+        &self,
+        title: String,
+        wordpress_menu_id: i64,
+        slug: String) -> Result<String>{
+        query_file!(
+            "queries/upsert_collection.sql",
+            slug,
+            title,
+            wordpress_menu_id)
+            .execute(&self.client)
+            .await?;
+        Ok(slug)
+    }
+
+    pub async fn insert_all_chapters(
+        &self,
+        chapters: Vec<Chapter>,
+        slug: String) -> Result<String>{
+
+        let mut tx = self.client.begin().await?;
+
+        // Delete all chapter data before re-inserting
+        query_file!(
+            "queries/delete_chapters.sql")
+        .execute(&mut tx)
+        .await?;
+        
+        let mut chapter_stack = Vec::new();  
+        let initial_tuple = (0, slug.clone());
+        chapter_stack.push(initial_tuple);
+
+        for current_chapter in chapters {
+
+            // Get chapter Doc ID from document title --> do in query
+            let chapter_doc_name = current_chapter.document_title;
+            
+            // Use stack to build chapter slug
+            let mut before_chapter_index = chapter_stack.last().clone().unwrap().0;
+            while before_chapter_index != (current_chapter.index_in_parent - 1)  {
+                let last_chapter = chapter_stack.last().unwrap();
+                let last_chapter_index = last_chapter.0;
+                if last_chapter_index >= current_chapter.index_in_parent {
+                    chapter_stack.pop();
+                }  
+                before_chapter_index = chapter_stack.last().clone().unwrap().0; 
+            }
+
+            // Concatenate URL Slugs of all chapters on the path
+            let mut chapter_stack_cur = chapter_stack.clone();
+            let mut url_slug_cur = current_chapter.url_slug.clone();
+            let final_path = (current_chapter.index_in_parent, current_chapter.url_slug.clone());
+            chapter_stack.push(final_path);
+
+            while !(chapter_stack_cur.is_empty()){
+                let temp_chapter = chapter_stack_cur.pop().unwrap();
+                let cur_slug = temp_chapter.1;
+                url_slug_cur = cur_slug + "." + &url_slug_cur;
+            }
+
+             // Insert chapter data into database
+             query_file!(
+                 "queries/insert_all_chapters.sql",
```

Comment:
This is called `insert_all_chapters` but it appears to only insert one chapter. There are a bunch of queries that insert multiple rows, it's a bit weird to read but to find them just search for `unnest`. On the rust end, that means passing one argument per field where each is a list. So instead of this:
```
insert(a1, b1, c1);
insert(a2, b2, c2);
```

You are doing this:
```
insert_all([a1, a2], [b1, b2], [c1, c2])
```

Reply:
Could I just rename the query I wrote so it's not misleading, or should I make a new query which uses unnest?

Reply:
I haven't checked yet if you addressed this already, but for now you could rename it since we aren't inserting all that many chapters yet.